Add output intermediate mesh option to MeshGenerator

MengnanLi91 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

This PR #23486 add a new option to MeshGenerator so that the users can output and check any intermediate mesh object in `[Mesh]` block.

## Reason
<!--Why do you need this feature or what is the enhancement?-->

This feature allows the user to output and check the intermediate mesh object besides the final mesh file. It is useful for debugging and complex mesh design #23486

## Design
<!--A concise description (design) of the enhancement.-->

A new bool option `output` is introduced to `[Mesh]` block. It can be set _true_ in one mesh object to output the corresponding mesh file #23486

## Impact
<!--Will the enhancement change existing APIs or add something new?-->

A new option is added to MeshGenerator to allow users output and check any intermediate mesh object #23486

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

MengnanLi91 committed:
Basic function of intermediate mesh storage #23486

MengnanLi91 committed:
Add test case and doco

Update MeshGenerator.C

MengnanLi91 committed:
Update MeshGenerator.md

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1362650/) on 508b39d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23486/site)

This comment will be updated on new commits.

MengnanLi91 committed:
Add gold mesh files

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1364077/) on a621191 wanted to post the following:


## Framework coverage

No coverage report(s) were found for the base commit b966571c58.
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23486/coverage/framework)


## Modules coverage

No coverage report(s) were found for the base commit b966571c58.

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23486/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Conda (Rocky)](https://civet.inl.gov/job/1366287/) on c86e129 : invalidated by @MengnanLi91

moosebuild commented:
Job [Conda (Intel Mac)](https://civet.inl.gov/job/1366286/) on c86e129 : invalidated by @MengnanLi91

MengnanLi91 committed:
Add _save_in_meshes, remove _final_generated_meshes(in progress)

Co-authored-by: Logan Harbour<logan.harbour@inl.gov>

Working progress

Working progress

Add _save_in_meshes

loganharbour commented:
Let's make sure that we have separate PRs for the exodus/nemesis output and then the save in capability, to make the review easier

MengnanLi91 committed:
Add save meshes capability to a seperate PR, revert the commits in this PR

This reverts commit 534e0a850fafbbce0a4c35c3ce76bb15122b653c.

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1383504/) on c890746 : invalidated by @MengnanLi91

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1383504/) on c890746 : invalidated by @MengnanLi91

loganharbour commented on framework/doc/content/source/meshgenerators/MeshGenerator.md:
````diff
@@ -142,3 +142,41 @@ GeneratedMeshGenerator 'gmg':    Bounding box maximum: (x,y,z)=(       1,
 GeneratedMeshGenerator 'gmg':   Global mesh volume = 1
 ```
+
+The parameter [!param](/Mesh/GeneratedMeshGenerator/output) output a mesh file immediately after it is generated. It can be used to view the intermediate mesh block during mesh generation.
````
```suggestion
The parameter [!param](/Mesh/GeneratedMeshGenerator/output) enables the output of a mesh file immediately after it is generated. It can be used to view the intermediate meshes during mesh generation.
```

loganharbour commented on framework/doc/content/source/meshgenerators/MeshGenerator.md:
````diff
@@ -142,3 +142,41 @@ GeneratedMeshGenerator 'gmg':    Bounding box maximum: (x,y,z)=(       1,
+
+```
+[Mesh]
+  [./left]
+    type = GeneratedMeshGenerator
+    dim = 2
+    nx = 3
+    ny = 3
+    xmin = -3
+    xmax = 0
+    ymin = -5
+    ymax = 5
+    output = true
+    nemesis = true
+  [../]
+  [./right]
+    type = GeneratedMeshGenerator
+    dim = 2
+    nx = 3
+    ny = 3
+    xmin = 3
+    xmax = 6
+    ymin = -5
+    ymax = 5
+  [../]
+
+  [./left_and_right]
+    type = MeshCollectionGenerator
+    inputs = 'left right'
+  [../]
````
Let's include the input file from the test instead so that they're linked

loganharbour commented on framework/doc/content/source/meshgenerators/MeshGenerator.md:
````diff
@@ -142,3 +142,41 @@ GeneratedMeshGenerator 'gmg':    Bounding box maximum: (x,y,z)=(       1,
+[]
+```
+
+The above will result in an intermediate mesh file 'left_in.e' besides the final mesh file 'output_intermediate_mesh_in.e'. This feature works for both REPLICATED and DISTRIBUTED mesh type.
````
```suggestion
The above will result in an intermediate mesh file 'left_in.e' in addition to the final mesh file 'output_intermediate_mesh_in.e' when ran in mesh only mode.
```

We need to chat about the replicated/distributed case. I haven't decided if we want nemesis output by default. Some people will run in parallel, but then want to use the output mesh as a serialized mesh for input and then this won't work. I think we may need to provide the option to output distributed as exodus.

loganharbour commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -11,6 +11,12 @@
+
+#include "libmesh/exodusII_io.h"
+#include "libmesh/checkpoint_io.h"
+#include <libmesh/nemesis_io.h>
```
```suggestion
#include "libmesh/nemesis_io.h"
```

loganharbour commented on test/tests/meshgenerators/output_intermediate_mesh/output_intermediate_mesh.i:
```diff
@@ -0,0 +1,34 @@
+    ymin = -5
+    ymax = 5
+    output = true
+    nemesis = true
```
This parameter is unused

loganharbour commented on test/tests/meshgenerators/output_intermediate_mesh/output_intermediate_mesh.i:
```diff
@@ -0,0 +1,34 @@
+[Mesh]
+  [./left]
```
We should enable moose input formatting in your vs code. This `[./xxx]` and `[../]` syntax is old and we don't use it anymore. The input formatter will do this for you.

loganharbour commented on test/tests/meshgenerators/output_intermediate_mesh/tests:
```diff
@@ -0,0 +1,31 @@
+  [intermediate_mesh_output]
+  design = 'MeshGenerator.md'
+  issues = '#23486'
+  requirement = 'The system shall be capable of output and check the intermediate mesh object besides the final mesh file'
```
These requirements should be as general as possible. The detail here is too specific. While we are outputting only the left block, the _purpose_ of the test is to test this with replicated mesh. How about:

The system shall support the file output of intermediate generated meshes in:
- a serialized form
- a distributed form

MengnanLi91 committed:
Address review: improve doco and test

Co-Authored-By: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1394447/) on 2dfad8a : invalidated by @MengnanLi91

MengnanLi91 committed:
Update test: only output nemesis mesh when requested

loganharbour commented on test/tests/meshgenerators/output_intermediate_mesh/tests:
```diff
@@ -0,0 +1,31 @@
+  issues = '#23486'
+  requirement = 'The system shall support the file output of intermediate generated meshes in'
+
+    [REPLICATED]
+        type = 'Exodiff'
+        input = 'output_intermediate_mesh.i'
+        exodiff = 'left_in.e'
+        cli_args = "--mesh-only"
+        recover = false
+        mesh_mode = 'REPLICATED'
+
+        detail = "Output only the left mesh block in a serialized mesh form"
+    []
+    [DISTRIBUTED]
+      type = 'Exodiff'
+      input = 'output_intermediate_mesh.i'
+      exodiff = 'left_in.e.2.0 left_in.e.2.1'
+      cli_args = "Mesh/left/parallel_type=distributed Mesh/left/nemesis=true --mesh-only"
+      recover = false
+      mesh_mode = 'DISTRIBUTED'
+      min_parallel = 2
+      max_parallel = 2
+
+      detail = "Output only the left mesh block in a distributed mesh form"
+    []
+
```
```suggestion
    [exodus]
        type = 'Exodiff'
        input = 'output_intermediate_mesh.i'
        exodiff = 'left_in.e'
        cli_args = "--mesh-only"
        recover = false

        detail = "the exodus file format"
    []
    [nemesis]
      type = 'Exodiff'
      input = 'output_intermediate_mesh.i'
      exodiff = 'left_in.e.2.0 left_in.e.2.1'
      cli_args = "Mesh/left/nemesis=true --mesh-only"
      recover = false
      min_parallel = 2
      max_parallel = 2
      detail = "the nemesis file format"
    []
```

loganharbour commented on test/tests/meshgenerators/output_intermediate_mesh/output_intermediate_mesh.i:
```diff
@@ -0,0 +1,33 @@
+    inputs = 'left right'
+  []
+[]
+
+[Outputs]
+  exodus = true
+[]
```
```suggestion
```

loganharbour commented on framework/src/meshgenerators/MeshGenerator.C:
```diff
@@ -21,6 +27,10 @@ MeshGenerator::validParams()
+  params.addParam<bool>(
+      "output", false, "Whether or not to output the mesh file after generating the mesh");
+  params.addParam<bool>(
+      "nemesis", false, "Whether or not to output the mesh file in distributed form");
```
```suggestion
      "nemesis", false, "Whether or not to output the mesh file in the nemesis format (only if output = true)");
```

You'll probably want to do this one on your computer, it'll probably fail clang format

loganharbour reviewed (changes requested):
Good after this!

MengnanLi91 committed:
Address review: exodus test should work for both replicated and distributed mesh

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1397072/) on 3e3bb6f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23486/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23486/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 571e078e34953f4a725ac28d7bf9e564c53eb9ca`

MengnanLi91 committed:
Clean format for Precheck

MengnanLi91 committed:
Update nemesis files in gold folder

MengnanLi91 committed:
Comment out nemesis test for now

The nemesis test is commented for now and wait for next libmesh update to make 'nemesis_io.set_hdf5_writing(false)'  available

loganharbour reviewed (changes requested):
Please add a comment above the commented out test as to why we're not testing this, and remove the nemesis gold files because they're not used

MengnanLi91 committed:
Address review: add comments for nemesis test, remove nemesis gold file
